File: apps/node/src/group-send.ts
import {
  canonical,
  createBundle,
  decryptBundle,
  verifyBundle,
  type Identity,
} from "../../../packages/core/src/index.js";
import type { ProfileDatabase } from "../../../packages/profile/src/database.js";
import {
  readProfileState,
  writeProfileState,
} from "../../../packages/profile/src/state.js";
import { GroupLedger } from "../../../packages/groups/src/ledger.js";
import { parseGroupBinding } from "../../../packages/groups/src/access.js";
import type { Content } from "./node.js";
import type { PrivateState } from "./local-state.js";
import { validateContent } from "./content-validation.js";
import {
  admitOutbox,
  requireOperationId,
  outboxPreview,
  type OutboxEntry,
} from "./outbox.js";
import { reconcileGroupOutbox } from "./group-outbox.js";

/** Locally authored bytes are signed and fully verified here. The admission
 * and preparing intent commit together before the content store or network is
 * touched. This never accepts an HTTP-supplied historical context or summary.
 * A missing payload after a crash is not reconstructed from the preview. */
export function commitGroupSendIntent(
  database: ProfileDatabase,
  identity: Identity,
  state: PrivateState,
  privateDigest: string,
  operationId: string,
  fingerprint: string,
  input: Content,
  recipients: string[],
  ttlMs: number,
  blocked: readonly string[],
  protectedIds: ReadonlySet<string>,
) {
  requireOperationId(operationId);
  validateContent(input);
  const binding = parseGroupBinding(input);
  if (
    input.type !== "message" ||
    !binding.groupEpoch ||
    binding.groupAudience === "historical"
  )
    throw new Error("Esta operação exige uma mensagem de grupo");
  if (!input.text?.trim() && !input.attachments?.length)
    throw new Error("Escreva uma mensagem ou junte um anexo");
  const groupEpoch = binding.groupEpoch;
  const next = structuredClone(state);
  return database.transaction((tx) => {
    if (readProfileState(tx)?.digest !== privateDigest)
      throw new Error("O estado privado mudou antes do envio");
    return GroupLedger.run(tx, identity, (ledger, registry) => {
      reconcileGroupOutbox(ledger, next.outbox);
      if (next.outbox?.[operationId])
        throw new Error("A intenção já existe; consulte o seu estado");
      const current = registry.state(binding.conversation);
      if (current.status !== "active" || current.head?.id !== groupEpoch)
        throw new Error(
          "O grupo mudou. Reveja os destinatários antes de enviar.",
        );
      const snapshot = registry.privateState(binding.conversation, groupEpoch);
      if (!snapshot)
        throw new Error("Falta confirmar o estado privado do grupo");
      let cards = snapshot.members;
      if (binding.groupAudience === "target") {
        const target = ledger.accepted(input.replyTo!);
        if (
          !target ||
          target.context.groupId !== binding.conversation ||
          target.context.epochId !== binding.targetEpoch ||
          target.context.kind !== "message"
        )
          throw new Error(
            "A resposta exige uma mensagem de grupo já verificada",
          );
        cards = cards.filter((card) =>
          target.context.readers.includes(card.id),
        );
      }
      const ids = cards.map((card) => card.id).sort();
      if (
        ids.length < 2 ||
        !ids.includes(identity.public.id) ||
        canonical(ids) !==
          canonical([...new Set([...recipients, identity.public.id])].sort())
      )
        throw new Error(
          "Os destinatários não correspondem à audiência verificada do grupo",
        );
      if (ids.some((id) => blocked.includes(id)))
        throw new Error("Há um destinatário bloqueado");
      if (
        input.members !== undefined &&
        canonical(input.members) !== canonical(cards)
      )
        throw new Error(
          "Os cartões indicados não correspondem ao grupo verificado",
        );
      const content = { ...structuredClone(input), members: cards };
      validateContent(content);
      const bundle = createBundle(identity, "message", content, cards, ttlMs);
      verifyBundle(bundle);
      const verified = decryptBundle(bundle, identity) as Content;
      validateContent(verified);
      if (canonical(verified) !== canonical(content))
        throw new Error("O conteúdo assinado não corresponde ao envio");
      const entry: OutboxEntry = {
        operationId,
        fingerprint,
        id: bundle.manifest.id,
        author: identity.public.id,
        conversation: binding.conversation,
        groupEpoch: groupEpoch,
        groupStopped: false,
        preview: outboxPreview(
          content.text || content.attachments?.[0]?.name || "Mensagem",
        ),
        created: bundle.manifest.created,
        expires: bundle.manifest.expires,
        priority:
          content.priority ?? (content.attachments?.length ? "bulk" : "normal"),
        bytes: Buffer.byteLength(canonical(bundle)),
        phase: "preparing",
        attempts: 0,
        lastAttemptAt: 0,
        nextAttemptAt: 0,
        lastError: "",
        manualPin: false,
        confirmations: Object.fromEntries(
          ids.filter((id) => id !== identity.public.id).map((id) => [id, {}]),
        ),
      };
      next.outbox = admitOutbox(next.outbox ?? {}, entry, Date.now());
      const admission = ledger.consider(
        {
          id: entry.id,
          kind: "message",
          author: bundle.manifest.author,
          readers: bundle.manifest.keys.map((key) => key.reader),
          public: false,
          content: verified,
        },
        entry.expires,
        entry.bytes,
        new Set([
          ...protectedIds,
          ...Object.values(next.outbox).map((e) => e.id),
        ]),
      );
      if (admission.decision.status !== "accepted")
        throw new Error("A mensagem não pôde ser admitida neste grupo");
      ledger.retireStops(new Set(Object.keys(next.outbox)));
      const digest = writeProfileState(
        tx,
        Buffer.from(canonical(next)),
        privateDigest,
      );
      return { bundle, state: next, digest };
    }).value;
  });
}

File: native/app/group_send.go
package app

import (
	"bytes"
	"errors"
	"sort"
	"strings"
	"time"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupaccess"
	"github.com/JohnnyPBelo/relayloom/native/groupauthority"
	"github.com/JohnnyPBelo/relayloom/native/groupledger"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"github.com/JohnnyPBelo/relayloom/native/profilestate"
	"github.com/JohnnyPBelo/relayloom/native/transport"
)

func groupMessagePreviewSource(content Content) string {
	value := text(content["text"])
	if value == "" {
		if attachments, ok := content["attachments"].([]any); ok && len(attachments) > 0 {
			if file, err := object(attachments[0]); err == nil {
				value = text(file["name"])
			}
		}
	}
	if value == "" {
		value = "Mensagem"
	}
	return value
}

func groupMessagePreview(content Content) string {
	return outboxPreview(groupMessagePreviewSource(content))
}

// A locally authored bundle is fully verified before its admission and intent
// share one SQL commit. No content-store or network write happens before that
// commit. A preparing record whose payload is lost never signs a replacement.
func (n *Node) sendGroupLocked(operation, fingerprint string, input Content, recipients []string, ttl int64) (any, error) {
	if err := n.requireGroupReplayLocked(); err != nil {
		return nil, err
	}
	if err := validateContent(input); err != nil {
		return nil, err
	}
	binding, err := groupaccess.ParseBinding(input)
	if err != nil {
		return nil, err
	}
	if text(input["type"]) != "message" || binding.Epoch == "" || binding.Audience == "historical" {
		return nil, errors.New("esta operação exige uma mensagem de grupo")
	}
	attachments, _ := input["attachments"].([]any)
	if strings.TrimSpace(text(input["text"])) == "" && len(attachments) == 0 {
		return nil, errors.New("escreva uma mensagem ou junte um anexo")
	}
	next, err := copyPrivate(n.private, n.identity.Public.ID)
	if err != nil {
		return nil, err
	}
	digest := n.privateDigest
	var bundle core.Bundle
	var entry OutboxRecord
	update := n.updateGroupState
	if update == nil {
		update = n.privateDatabase.Update
	}
	protected := n.protectedGroupContentLocked()
	err = update(func(tx *groupstore.Tx) error {
		current, err := profilestate.Read(tx)
		if err != nil {
			return err
		}
		if current == nil || current.Digest != n.privateDigest {
			return errors.New("o estado privado mudou antes do envio")
		}
		_, err = groupledger.Run(tx, *n.identity, func(l *groupledger.Ledger, g *groupauthority.Registry) error {
			if _, err := reconcileGroupOutbox(l, next.Outbox); err != nil {
				return err
			}
			if _, ok := next.Outbox[operation]; ok {
				return errors.New("a intenção já existe; consulte o seu estado")
			}
			current, err := g.State(binding.Conversation)
			if err != nil {
				return err
			}
			if current.Status != "active" || current.Head == nil || current.Head.ID != binding.Epoch {
				return errors.New("o grupo mudou; reveja os destinatários antes de enviar")
			}
			snapshot, err := g.PrivateState(binding.Conversation, binding.Epoch)
			if err != nil {
				return err
			}
			if snapshot == nil {
				return errors.New("falta confirmar o estado privado do grupo")
			}
			cards := append([]core.PublicIdentity{}, snapshot.Members...)
			if binding.Audience == "target" {
				target, err := l.Accepted(text(input["replyTo"]))
				if err != nil {
					return err
				}
				if target == nil || target.Context.GroupID != binding.Conversation || target.Context.EpochID != binding.TargetEpoch || target.Context.Kind != "message" {
					return errors.New("a resposta exige uma mensagem de grupo já verificada")
				}
				filtered := []core.PublicIdentity{}
				for _, card := range cards {
					if contains(target.Context.Readers, card.ID) {
						filtered = append(filtered, card)
					}
				}
				cards = filtered
			}
			ids := sorted(memberIDs(cards))
			requested := map[string]bool{n.identity.Public.ID: true}
			for _, id := range recipients {
				requested[id] = true
			}
			selected := []string{}
			for id := range requested {
				selected = append(selected, id)
			}
			if len(ids) < 2 || !contains(ids, n.identity.Public.ID) || !equalIDs(ids, selected) {
				return errors.New("os destinatários não correspondem à audiência verificada do grupo")
			}
			for _, id := range ids {
				if contains(n.config.Blocked, id) {
					return errors.New("há um destinatário bloqueado")
				}
			}
			if supplied, exists := input["members"]; exists {
				a, err := core.Canonical(supplied)
				if err != nil {
					return err
				}
				b, err := core.Canonical(cards)
				if err != nil {
					return err
				}
				if !bytes.Equal(a, b) {
					return errors.New("os cartões indicados não correspondem ao grupo verificado")
				}
			}
			raw, err := cloneValue(input)
			if err != nil {
				return err
			}
			values, err := object(raw)
			if err != nil {
				return err
			}
			content := Content(values)
			content["members"] = cards
			if err := validateContent(content); err != nil {
				return err
			}
			bundle, err = core.CreateBundle(*n.identity, "message", content, cards, false, ttl)
			if err != nil {
				return err
			}
			if err := core.VerifyBundle(bundle); err != nil {
				return err
			}
			decoded, err := core.DecryptBundle(bundle, n.identity)
			if err != nil {
				return err
			}
			verified, err := object(decoded)
			if err != nil {
				return err
			}
			if err := validateContent(Content(verified)); err != nil {
				return err
			}
			a, err := core.Canonical(content)
			if err != nil {
				return err
			}
			b, err := core.Canonical(verified)
			if err != nil {
				return err
			}
			if !bytes.Equal(a, b) {
				return errors.New("o conteúdo assinado não corresponde ao envio")
			}
			encoded, err := core.Canonical(bundle)
			if err != nil {
				return err
			}
			priority := transport.Priority(text(content["priority"]))
			if priority == "" {
				priority = transport.Normal
				if len(attachments) > 0 {
					priority = transport.Bulk
				}
			}
			stopped := false
			entry = OutboxRecord{OperationID: operation, Fingerprint: fingerprint, ID: bundle.Manifest.ID, Author: n.identity.Public.ID, Conversation: binding.Conversation,
				GroupEpoch: binding.Epoch, GroupStopped: &stopped, Preview: groupMessagePreview(Content(verified)), Created: bundle.Manifest.Created, Expires: bundle.Manifest.Expires,
				Priority: priority, Bytes: int64(len(encoded)), Phase: "preparing", Confirmations: map[string]Confirmation{}}
			for _, id := range ids {
				if id != entry.Author {
					entry.Confirmations[id] = Confirmation{}
				}
			}
			now := time.Now().UnixMilli()
			pending, totalBytes := 0, entry.Bytes
			terminal := []OutboxRecord{}
			for _, record := range next.Outbox {
				if pendingOutbox(record, now) {
					pending++
					totalBytes += record.Bytes
				} else {
					terminal = append(terminal, record)
				}
			}
			if pending >= outboxPendingLimit || totalBytes > outboxBytesLimit {
				return errors.New("reserva de envios pendentes atingida; aguarde a recepção ou expiração")
			}
			sort.Slice(terminal, func(i, j int) bool {
				if terminal[i].Created == terminal[j].Created {
					return terminal[i].ID < terminal[j].ID
				}
				return terminal[i].Created < terminal[j].Created
			})
			for len(next.Outbox) >= outboxTotalLimit && len(terminal) > 0 {
				delete(next.Outbox, terminal[0].OperationID)
				terminal = terminal[1:]
			}
			if len(next.Outbox) >= outboxTotalLimit {
				return errors.New("limite do diário de envios atingido")
			}
			next.Outbox[operation] = entry
			retained := map[string]bool{}
			for operation, record := range next.Outbox {
				retained[operation] = true
				protected[record.ID] = true
			}
			admission, err := l.Consider(groupaccess.Candidate{ID: entry.ID, Kind: "message", Author: bundle.Manifest.Author, Readers: ids, Content: verified}, entry.Expires, entry.Bytes, protected, now)
			if err != nil {
				return err
			}
			if admission.Decision.Status != "accepted" {
				return errors.New("a mensagem não pôde ser admitida neste grupo")
			}
			if err := l.RetireStops(retained); err != nil {
				return err
			}
			data, err := core.Canonical(next)
			if err != nil {
				return err
			}
			digest, err = profilestate.Write(tx, data, &n.privateDigest)
			return err
		})
		return err
	})
	if err != nil {
		n.recoverGroupContentLocked()
		return nil, err
	}
	n.private, n.privateDigest = next, digest
	if _, err := n.Store.PutReserved(bundle, false); err != nil {
		failed, copyErr := copyPrivate(n.private, n.identity.Public.ID)
		if copyErr == nil {
			entry.Phase = "unavailable"
			entry.LastError = "Não foi possível guardar o conteúdo do envio. Não foi criado outro ID."
			failed.Outbox[operation] = entry
			copyErr = n.persistPrivateLocked(failed)
		}
		if copyErr == nil {
			copyErr = n.restoreGroupHoldsLocked()
		}
		if copyErr != nil {
			_ = n.lockPrivateLocked()
		}
		return nil, err
	}
	next, err = copyPrivate(n.private, n.identity.Public.ID)
	if err != nil {
		return nil, err
	}
	entry.Phase = "ready"
	next.Outbox[operation] = entry
	if err := n.persistPrivateLocked(next); err != nil {
		return nil, err
	}
	if n.connectedOutboxLocked() {
		if err := n.attemptOutboxLocked(operation, time.Now().UnixMilli()); err != nil && n.identity == nil {
			return nil, err
		}
	}
	return n.outboxResponseLocked(operation)
}
